core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -104,19 +119,40 @@ public void run() {
   /**
    * Transfers all pending and future requests and mutations to the given stream.
    *
-   * <p>No-op if either this method or {@link #cancel} have already been called.
+   * <p>This method must be called at most once.  Extraneous calls will throw and end up cancelling
+   * the given streams.
+   *
+   * <p>If {@link #cancel} has been called, this method will cancel the given stream.
    */
   // When this method returns, passThrough is guaranteed to be true
   final void setStream(ClientStream stream) {
+    ClientStream savedRealStream;
+    Status savedError;
     synchronized (this) {
-      // If realStream != null, then either setStream() or cancel() has been called.
-      if (realStream != null) {
-        return;
+      savedRealStream = realStream;
+      savedError = error;
+      if (savedRealStream == null) {
+        realStream = checkNotNull(stream, "stream");
       }
-      realStream = checkNotNull(stream, "stream");
     }
 
-    drainPendingCalls();
+    if (savedRealStream == null) {
+      drainPendingCalls();
+    } else {
+      // If realStream was not null, then either setStream() or cancel() must had been called,
+      // we will cancel and discard the given stream.
+      // ClientStream.cancel() must be called after start()
+      stream.start(NOOP_STREAM_LISTENER);
+      if (savedError != null) {
+        stream.cancel(savedError);
```

Comment:
For my own record, the issue is `setStream()` may be scheduled for an unbounded period of time into the future. For example, `CallCredentials` may take a long time to fetch the credentials before triggering `setStream()` (in `MetadataApplierImpl`, prior to which the RPC may be cancelled due to deadline-exceeded. In those cases, real stream are created before `setStream()` is called, only to be started (as required by cancel()) and then cancelled. `start()` typically involve I/O, e.g., sending headers, which are wasted. If the delay before `setStream()` is long enough, this will have a high chance to happen.